integration/spark/src/main/scala/org/apache/carbondata/trash/DataTrashManager.scala

```diff
@@ -0,0 +1,168 @@
+   * move stale segment to trash folder, but not include compaction segment
+   */
+  private def moveStaleSegmentsToTrash(carbonTable: CarbonTable): Unit = {
+    if (carbonTable.isHivePartitionTable) {
+      CleanFilesUtil.cleanStaleSegmentsForPartitionTable(carbonTable)
+    } else {
+      CleanFilesUtil.cleanStaleSegments(carbonTable)
+    }
+  }
+
+  private def cleanExpiredSegments(
+      carbonTable: CarbonTable,
+      isForceDelete: Boolean,
+      cleanStaleInProgress: Boolean,
+      partitionSpecsOption: Option[Seq[PartitionSpec]]): Unit = {
+    val partitionSpecs = partitionSpecsOption.map(_.asJava).orNull
+    SegmentStatusManager.deleteLoadsAndUpdateMetadata(carbonTable,
+      isForceDelete, partitionSpecs, cleanStaleInProgress, true)
+    if (carbonTable.isHivePartitionTable && partitionSpecsOption.isDefined) {
+      SegmentFileStore.cleanSegments(carbonTable, partitionSpecs, isForceDelete)
+    }
+  }
+
+  /**
+   * clean the stale compact segment immediately after compaction failure
+   */
+  def cleanStaleCompactionSegment(
+      carbonTable: CarbonTable,
+      mergedLoadName: String,
+      factTimestamp: Long,
+      partitionSpecs: Option[Seq[PartitionSpec]]): Unit = {
+    val metadataFolderPath = CarbonTablePath.getMetadataPath(carbonTable.getTablePath)
+    val details = SegmentStatusManager.readLoadMetadata(metadataFolderPath)
+    if (details == null || details.isEmpty) {
+      return
+    }
+    val loadDetail = details.find(detail => mergedLoadName.equals(detail.getLoadName))
+    // only clean stale compaction segment
+    if (loadDetail.isEmpty) {
+      val segmentId = mergedLoadName.split(CarbonCommonConstants.UNDERSCORE)(1)
+      if (carbonTable.isHivePartitionTable) {
+        if (partitionSpecs.isDefined) {
+          partitionSpecs.get.foreach { partitionSpec =>
+            cleanStaleCompactionDataFiles(
+              partitionSpec.getLocation.toString, segmentId, factTimestamp)
+          }
+        }
+      } else {
+        val segmentPath = CarbonTablePath.getSegmentPath(carbonTable.getTablePath, segmentId)
+        cleanStaleCompactionDataFiles(
+          segmentPath, segmentId, factTimestamp)
+      }
+    }
+  }
+
+  private def cleanStaleCompactionDataFiles(
+      folderPath: String,
+      segmentId: String,
+      factTimestamp: Long): Unit = {
+    if (FileFactory.isFileExist(folderPath)) {
+      val namePart = CarbonCommonConstants.HYPHEN + segmentId +
+        CarbonCommonConstants.HYPHEN + factTimestamp
+      val toBeDelete = FileFactory.getCarbonFile(folderPath).listFiles(new CarbonFileFilter() {
+        override def accept(file: CarbonFile): Boolean = {
+          file.getName.contains(namePart)
+        }
+      })
+      if (toBeDelete != null && toBeDelete.nonEmpty) {
+        try {
+          CarbonUtil.deleteFoldersAndFilesSilent(toBeDelete: _*)
+        } catch {
+          case e: Throwable =>
+            LOGGER.error("Exception in deleting the delta files." + e)
```

Comment:
delta files? Can we change this and also include the segmentID too?

Reply:
maybe instead of cleaning compacted segments after some compaction failure, we can maybe try and clear before the compaction itself. Let's say we are trying to compact to segment 0.1, so there we can.
```1. If segment 0.1 already exists -> clean it -> If the cleaning fails -> fail the compaction command itself ```
```2. else -> proceed with compaction```

In the current implementation, if cleaning the stale compact segment fails, and if the user does not manually clean it, there could be issues in the next compaction command. If we do it before compaction we can just fail the compaction itself and ask user to clean manually. What do you think?

Reply:
fixed the exception message.
clean data before compaction is a good suggestion, but it may remove data files(which create by other jobs) by mistake.